src/main/java/de/zalando/aruha/nakadi/service/EventStream.java

```diff
@@ -55,23 +88,37 @@ public void streamEvents() {
             int messagesRead = 0;
             int messagesReadInBatch = 0;
 
-            Map<String, List<String>> currentBatch = emptyBatch();
-            final Map<String, String> latestOffsets = Maps.newHashMap(config.getCursors());
+            Map<TopicPartition, List<String>> currentBatch = emptyBatch();
+            Map<TopicPartition, String> latestOffsets = cursorsToOffsetMap(initialCursors);
 
             long start = currentTimeMillis();
             long batchStartTime = start;
 
             while (true) {
+
+                // check if rebalance is invoked
+                final Optional<List<Cursor>> newCursorsOrNone = newDistribution.getAndSet(Optional.empty());
+                if (newCursorsOrNone.isPresent()) {
+                    final List<Cursor> newCursors = newCursorsOrNone.get();
+                    eventConsumer.setCursors(newCursors);
+                    partitions = cursorsToPartitions(newCursors);
+                    latestOffsets = cursorsToOffsetMap(newCursors);
+                    // forget about all messages not flushed
+                    messagesReadInBatch = 0;
+                    currentBatch = emptyBatch();
+                    batchStartTime = currentTimeMillis();
+                }
+
                 final Optional<ConsumedEvent> eventOrEmpty = eventConsumer.readEvent();
```

Comment:
ah ok I see, that will make reading from kafka to be in-sync with batch timeout, right?